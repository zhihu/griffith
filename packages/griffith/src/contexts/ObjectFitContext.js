import React from 'react'

const ObjectFitContext = React.createContext({})
ObjectFitContext.displayName = 'ObjectFitContext'

export default ObjectFitContext
