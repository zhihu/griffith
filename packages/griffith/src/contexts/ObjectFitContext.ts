import React from 'react'

type ObjectFit = 'fill' | 'contain' | 'cover' | 'none' | 'scale-down'
type ObjectFitContextValue = {
  objectFit?: ObjectFit
  setObjectFit?(objectFit: ObjectFit): void
}

const ObjectFitContext = React.createContext<ObjectFitContextValue>({})
ObjectFitContext.displayName = 'ObjectFitContext'

export default ObjectFitContext
