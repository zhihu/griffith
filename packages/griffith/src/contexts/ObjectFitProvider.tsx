import React, {useEffect, useMemo, useState} from 'react'
import ObjectFitContext, {ObjectFit} from './ObjectFitContext'

type Props = {
  initialObjectFit?: ObjectFit
}

const ObjectFitProvider: React.FC<Props> = ({
  initialObjectFit = 'contain',
  children,
}) => {
  const [objectFit, setObjectFit] = useState<ObjectFit>(initialObjectFit)
  const contextValue = useMemo(
    () => ({
      objectFit,
      setObjectFit,
    }),
    [objectFit, setObjectFit]
  )

  useEffect(() => {
    setObjectFit(initialObjectFit)
  }, [initialObjectFit])

  return (
    <ObjectFitContext.Provider value={contextValue}>
      {children}
    </ObjectFitContext.Provider>
  )
}

export default ObjectFitProvider
