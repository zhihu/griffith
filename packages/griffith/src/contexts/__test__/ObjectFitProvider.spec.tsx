import React, {useContext, useEffect} from 'react'
import {act, render, RenderResult} from '@testing-library/react'
import ObjectFitContext, {
  ObjectFit,
  ObjectFitContextValue,
} from '../ObjectFitContext'
import ObjectFitProvider from '../ObjectFitProvider'

test('ObjectFitProvider', () => {
  let ctx!: ObjectFitContextValue
  const values: ObjectFit[] = []
  const MyComponent = () => {
    const value = (ctx = useContext(ObjectFitContext))
    useEffect(() => {
      values.push(value.objectFit)
    }, [value])
    return null
  }

  let result!: RenderResult
  act(() => {
    result = render(
      <ObjectFitProvider>
        <MyComponent />
      </ObjectFitProvider>
    )
  })

  act(() => {
    ctx.setObjectFit('fill')
  })

  act(() => {
    result.rerender(
      <ObjectFitProvider initialObjectFit="cover">
        <MyComponent />
      </ObjectFitProvider>
    )
  })

  expect(values).toMatchInlineSnapshot(`
  Array [
    "contain",
    "fill",
    "cover",
  ]
  `)
})
