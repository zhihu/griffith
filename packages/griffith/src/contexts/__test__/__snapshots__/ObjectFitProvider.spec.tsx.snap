// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectFitProvider 1`] = `
[
  "contain",
  "fill",
  "cover",
]
`;
